sdk/core/azure-core-experimental/azure/core/experimental/transport/_urllib3.py

```diff
@@ -0,0 +1,300 @@
+# --------------------------------------------------------------------------
+#
+# Copyright (c) Microsoft Corporation. All rights reserved.
+#
+# The MIT License (MIT)
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the ""Software""), to
+# deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+# sell copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+# --------------------------------------------------------------------------
+import os
+import functools
+from typing import Any, Callable, Mapping, Optional, Union, Dict, cast, List, Tuple, Iterator
+
+import urllib3
+
+from azure.core.pipeline import Pipeline
+from azure.core.exceptions import ServiceRequestError, ServiceResponseError, IncompleteReadError, DecodeError
+from azure.core.configuration import ConnectionConfiguration
+from azure.core.pipeline.transport import HttpTransport
+from azure.core.rest import HttpRequest as RestHttpRequest, HttpResponse as RestHttpResponse
+from azure.core.rest._http_response_impl import HttpResponseImpl
+
+
+DEFAULT_BLOCK_SIZE = 32768
+_TYPE_FIELD_VALUE = Union[str, bytes]
+_TYPE_FIELD_VALUE_TUPLE = Union[
+    _TYPE_FIELD_VALUE,
+    Tuple[str, _TYPE_FIELD_VALUE],
+    Tuple[str, _TYPE_FIELD_VALUE, str],
+]
+
+
+def _read_files(fields, files):
+    if files is None:
+        return
+    if isinstance(files, Mapping):
+        files = list(files.items())
+    for k, v in files:
+        if hasattr(v, "read"):
+            updated = v.read()
+        elif isinstance(v, tuple):
+            updated = tuple(i.read() if hasattr(i, "read") else i for i in v)
```

Comment:
This method force-reads all multipart file data into memory before sending the request as `urllib3` does not support file-like objects here. And this also seems to be what `requests` was doing all along:
https://github.com/psf/requests/blob/main/src/requests/models.py#L190-L191